### PR TITLE
Experimentally add bounds checks to charCodeAt calls to avoid deoptimization

### PR DESCRIPTION
**Pull request by amcasey**
Noticed in #41482

### CODE REVIEW

**Comment by amcasey** [comment]
@typescript-bot perf test this

File: src/compiler/path.ts [hunk 1]
```diff
@@ -155,18 +156,19 @@ namespace ts {
...
-                        if (path.charCodeAt(volumeSeparatorEnd) === CharacterCodes.slash) {
-                            // URL: "file:///c:/", "file://localhost/c:/", "file:///c%3a/", "file://localhost/c%3a/"
-                            return ~(volumeSeparatorEnd + 1);
-                        }
                         if (volumeSeparatorEnd === path.length) {
                             // URL: "file:///c:", "file://localhost/c:", "file:///c$3a", "file://localhost/c%3a"
                             // but not "file:///c:d" or "file:///c%3ad"
                             return ~volumeSeparatorEnd;
                         }
+                        if (path.charCodeAt(volumeSeparatorEnd) === CharacterCodes.slash) {
```
Comment (DanielRosenwasser) [inline src/compiler/path.ts:168]: Why'd this move lower?

Comment (amcasey) [reply src/compiler/path.ts:168]: It seemed silly to do the length check after accessing the character that position.

Comment (DanielRosenwasser) [reply src/compiler/path.ts:168]: Gotcha. Looking at it more closely, I don't even know if `volumeSeparatorEnd === path.length` is even a valid condition, but I'm not familiar enough with this code.

**Comment by DanielRosenwasser** [comment]
Seems like there's no obvious improvement, and I can't tell if the slightly higher times are in the margin of error. If they are, then I think the change is worth it just to avoid any future noise when doing these investigations.

**Comment by amcasey** [comment]
@DanielRosenwasser I don't fully understand the trade-off between adding an explicit bounds check (extra work) and getting de-optimized.  I picked a few spots that either showed up in a v8 trace or seemed like low-hanging fruit (e.g. because one bounds check would cover several calls).

I agree that there is no obvious perf win and I don't think that's because of coverage issues.  Personally, I don't think these deopts stand out in a trace with hundreds, so I'm not sure the change reduces noise either.

Maybe @rbuckton or @weswigham has thoughts on whether this change is valuable in the absence of an empirical perf win?

**Comment by DanielRosenwasser** [comment]
> @DanielRosenwasser I don't fully understand the trade-off between adding an explicit bounds check (extra work)

It all depends on how often these things get used in a hot loop honestly. From my limited knowledge, that de-opt will mean that you can't rely on the character codes coming back as ["small integers"](https://github.com/v8/v8/blob/86991d0587a1f05b973eef8420c0cf48fdaaf915/src/objects/smi.h) in V8, and because they come back as `NaN`, the code will go through both a "bail-out" (the de-opt itself) and rewriting the code to use a ["heap number"](https://github.com/v8/v8/blob/86991d0587a1f05b973eef8420c0cf48fdaaf915/src/objects/heap-number.h) (which is a boxed double).

**Pull request closed without merging**